### PR TITLE
Feature: RPC-based state reader for Blockifier reexecution

### PR DESCRIPTION
**Pull request by odesenfans**
Problem: we need a way to reexecute blocks with Blockifier.

Solution: implement the `StateReader` trait using the `starknet-rs` crate to fetch all the relevant data from a Pathfinder node.

Issue Number: N/A

## Type

- [x] feature
- [ ] bugfix
- [ ] dev (no functional changes, no API changes)
- [ ] fmt (formatting, renaming)
- [ ] build
- [ ] docs
- [ ] testing

## Breaking changes?

- [ ] yes
- [x] no

### CODE REVIEW

File: crates/starknet-replay/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,17 @@
+[package]
+name = "starknet-replay"
```
Comment (notlesh) [inline crates/starknet-replay/Cargo.toml:2]: there are already 2 existing crates/projects called `starknet-replay`, I'd suggest not creating a 3rd ;)

Comment (odesenfans) [reply crates/starknet-replay/Cargo.toml:2]: I need a better name :(

File: crates/starknet-replay/src/rpc_state_reader.rs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+use starknet_api::core::{ClassHash, CompiledClassHash, ContractAddress, Nonce};
+use starknet_api::hash::StarkFelt;
+use starknet_api::state::StorageKey;
+use starknet_os_types::deprecated_compiled_class::GenericDeprecatedCompiledClass;
+use starknet_os_types::hash::GenericClassHash;
+use starknet_os_types::sierra_contract_class::GenericSierraContractClass;
+
+use crate::utils::{execute_coroutine, felt_api2vm, felt_vm2api};
+
+pub struct AsyncRpcStateReader<T>
```
Comment (notlesh) [inline crates/starknet-replay/src/rpc_state_reader.rs:16]: Is `T` here just for the `JsonRpcClient<T>`? We could give it the concrete type here of `HttpTransport` to make our lives easier unless we have a good reason to keep that generic.

Comment (notlesh) [reply crates/starknet-replay/src/rpc_state_reader.rs:16]: Meh, this wasn't a big deal, I wouldn't change it unless it makes sense

File: crates/starknet-replay/src/rpc_state_reader.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+fn to_state_err<E: ToString>(e: E) -> StateError {
+    StateError::StateReadError(e.to_string())
+}
+
+impl<T> AsyncRpcStateReader<T>
+where
+    T: JsonRpcTransport + Sync + Send + 'static,
+{
+    async fn get_storage_at_async(&self, contract_address: ContractAddress, key: StorageKey) -> StateResult<StarkFelt> {
```
Comment (notlesh) [inline crates/starknet-replay/src/rpc_state_reader.rs:45]: I would make these `pub` if the `Async` in `AsyncRpcStateReader` is to be publicly meaningful :)

File: crates/starknet-replay/Cargo.toml (hunk 1 above)
Comment (notlesh) [reply crates/starknet-replay/Cargo.toml:2]: maybe `rpc-replay`? not great, but at least it's unique

Comment (odesenfans) [reply crates/starknet-replay/Cargo.toml:2]: Sold!

File: crates/starknet-replay/src/rpc_state_reader.rs (hunk 3 above)
Comment (odesenfans) [reply crates/starknet-replay/src/rpc_state_reader.rs:45]: Makes sense!

**Review by notlesh — APPROVED** [review]

**Pull request merged**